samples/server.c

```diff
@@ -299,20 +319,47 @@ static size_t
 migration_read_data(vfu_ctx_t *vfu_ctx, void *buf, __u64 size, __u64 offset)
 {
     struct server_data *server_data = vfu_get_private(vfu_ctx);
+    uint8_t *p;
+    size_t bar_size;
 
     if (server_data->migration.data_size < size) {
         vfu_log(vfu_ctx, LOG_ERR, "invalid migration data read %#llx-%#llx",
                 offset, offset + size - 1);
         return -EINVAL;
     }
 
-    /* FIXME implement, client should be able to write any byte range */
-    assert((offset == 0 && size >= sizeof server_data->bar0)
-           || offset >= sizeof server_data->bar0);
+    /*
+     * If in pre-copy state we copy BAR1, if in stop-and-copy state we copy
+     * BAR0. This behavior is purely an artifact of this server implementation
+     * simply to make it as simple as possible. Note that the client might go
+     * from state running to stop-and-copy, completely skipping the pre-copy
+     * state. This is legitimate but we don't support it for now.
+     *
+     * FIXME implement transitioning from the running state straight to the
+     * stop-and-copy state.
+     */
 
-    if (offset == 0 && size >= sizeof server_data->bar0) {
-        memcpy(buf, &server_data->bar0, sizeof server_data->bar0);
+    if (server_data->migration.state == VFU_MIGR_STATE_PRE_COPY) {
+        p = server_data->bar1;
+        bar_size = server_data->bar1_size;
+    } else if (server_data->migration.state == VFU_MIGR_STATE_STOP_AND_COPY) {
+        p = (uint8_t*)&server_data->bar0;
+        bar_size = sizeof server_data->bar0;
+    } else {
+        /*
```

Comment:
add a log entry?

Reply:
We shouldn't call the device callaback in the first place, I'll add a FIXME in the library code.